ONSdigital/dp-api-router
# Feature/identity

### What

- Added zebedee health client, with health check for ZebedeeURL if audit is enabled
- Added identity check in audit middleware handler, calling `GET <zebedeeURL>/identity` to validate the identity
- Validated that the auth headers are propagated to the downstream service:
  - `X-Request-Id` (value obtained from context, as it might be newly generated)
  - `Collection-Id`
  - `Authorization`
  - `X-Florence-Token`
- Use`User-Identity`, or `Caller-Identity` if the first is not available, as audit event identity field.

### How to review

- Make sure that code changes make sense
- Make sure unit tests pass
- If you can run the following integration test (I did it already in my local machine):
  - run `make debug`
  - run the audit kafka consumer from https://github.com/ONSdigital/dp-data-tools/pull/10
  - run Zebedee or do an ssh tunnelling to access develop Zebedee instead.
  - Validate that the kafka audit messages are correct.
  - Optionally, run the downstream service to validate that the headers are propagated (e.g. dataset api)
Attached a postman collection to test the calls

### Who can review

Anyone


- Postman collection:

```
{
	"info": {
		"_postman_id": "5a98b8a7-aa50-4bb5-953e-73721c69ffe5",
		"name": "dp-api-router",
		"schema": "https://schema.getpostman.com/json/collection/v2.1.0/collection.json"
	},
	"item": [
		{
			"name": "GET datasets",
			"request": {
				"method": "GET",
				"header": [
					{
						"key": "Collection-Id",
						"value": "collection123",
						"type": "text"
					},
					{
						"key": "X-Request-Id",
						"value": "request666",
						"type": "text"
					},
					{
						"key": "User-Identity",
						"value": "identity321",
						"type": "text"
					},
					{
						"key": "X-Florence-Token",
						"value": "changeMe",
						"type": "text"
					},
					{
						"key": "Authorization",
						"value": "changeMe",
						"type": "text"
					}
				],
				"url": {
					"raw": "http://localhost:23200/v1/datasets?q1=v1&q2=v2",
					"protocol": "http",
					"host": [
						"localhost"
					],
					"port": "23200",
					"path": [
						"v1",
						"datasets"
					],
					"query": [
						{
							"key": "q1",
							"value": "v1"
						},
						{
							"key": "q2",
							"value": "v2"
						}
					]
				},
				"description": "Get datasets through dp-api-router with v1 and auth token"
			},
			"response": []
		},
		{
			"name": "Healthcheck",
			"request": {
				"method": "GET",
				"header": [],
				"url": {
					"raw": "http://localhost:23200/health",
					"protocol": "http",
					"host": [
						"localhost"
					],
					"port": "23200",
					"path": [
						"health"
					]
				}
			},
			"response": []
		}
	],
	"protocolProfileBehavior": {}
}
```
